File: server/src/routing/Router.cpp
#include "routing/Router.hpp"
#include "routing/Handlers.hpp"
#include "network/Session.hpp"

Router::Router(Handlers* handlers) : handlers_(handlers) {}
Json Router::route(const Json& request, std::shared_ptr<Session> session) {
    Json res;
    if(!request.contains("action")) {
        res["status"] = "error";
        res["message"] = "Missing action";
        if (request.contains("req_id")) res["req_id"] = request["req_id"].getString();
        return res;
    }
    std::string action = request["action"].getString();
    if(action == "register") res = handlers_->handleRegister(request);
    else if(action == "login") res = handlers_->handleLogin(request, session);
    else if(action == "send_message") res = handlers_->handleSendMessage(request);
    else if(action == "get_messages") res = handlers_->handleGetMessages(request);
    else if(action == "get_users") res = handlers_->handleGetUsers(request);
    else if(action == "logout") res = handlers_->handleLogout(request, session);
    else if(action == "upload_key") res = handlers_->handleUploadKey(request);
    else if(action == "get_key") res = handlers_->handleGetKey(request);
    else {
        res["status"] = "error";
        res["message"] = "Unknown action: " + action;
    }
    if(request.contains("req_id")) {
        res["req_id"] = request["req_id"].getString();
    }
    return res;
}
void Router::onUserDisconnected(const std::string& username) {
    if (handlers_) handlers_->userDisconnected(username);
}

File: server-python/neochat_server/router.py
"""
Router — dispatches a parsed request to the right Handlers method.

Mirrors src/routing/Router.hpp/cpp: looks at request["action"], calls the
matching handler, and echoes back request["req_id"] if the client sent
one (so clients can match async responses to requests).
"""
from __future__ import annotations

from typing import Optional

from .handlers import Handlers
from .session import Session

_NO_SESSION_ACTIONS = {"register", "send_message", "get_messages", "get_users", "upload_key", "get_key"}


class Router:
    def __init__(self, handlers: Handlers) -> None:
        self.handlers = handlers

    async def route(self, request: dict, session: Optional[Session]) -> dict:
        if "action" not in request:
            res = {"status": "error", "message": "Missing action"}
            if "req_id" in request:
                res["req_id"] = request["req_id"]
            return res

        action = str(request["action"])
        h = self.handlers

        if action == "register":
            res = await h.handle_register(request)
        elif action == "login":
            res = await h.handle_login(request, session)
        elif action == "send_message":
            res = await h.handle_send_message(request)
        elif action == "get_messages":
            res = await h.handle_get_messages(request)
        elif action == "get_users":
            res = await h.handle_get_users(request)
        elif action == "logout":
            res = await h.handle_logout(request, session)
        elif action == "upload_key":
            res = await h.handle_upload_key(request)
        elif action == "get_key":
            res = await h.handle_get_key(request)
        else:
            res = {"status": "error", "message": f"Unknown action: {action}"}

        if "req_id" in request:
            res["req_id"] = request["req_id"]
        return res

    def on_user_disconnected(self, username: str) -> None:
        self.handlers.user_disconnected(username)
